src/stdhasher.cr

```diff
@@ -0,0 +1,257 @@
+#       # mutate
+#       self
+#     end
+#
+#     def <<(v)
+#       # v.hash will mutate hasher
+#       v.hash(self)
+#       self
+#     end
+#
+#     # digest returns hashsum for current state without state mutation
+#     def digest : Value
+#     end
+#
+#     # should be implemented for `Hash#hash(hasher)`
+#     def clone_build
+#       with_state_copy do |copy|
+#         yield copy
+#         copy.digest
+#       end
+#     end
+#   end
+
+# StdHasher used as standard hasher in `Object#hash`
+# It have to provide defenense against HashDos, and be reasonably fast.
+# To protect against HashDos, it is seeded with secure random, and have
+# permutation that hard to forge without knowing seed and seeing hash digest.
+#
+# Also it has specialized methods for primitive keys with different seeds.
+struct StdHasher
+  alias Value = UInt32
+
+  @@seed = StaticArray(UInt32, 7).new { |i| 0_u32 }
+  buf = pointerof(@@seed).as(Pointer(UInt8))
+  SecureRandom.random_bytes(buf.to_slice(sizeof(typeof(@@seed))))
+
+  private def initialize(@h : Pointer(Impl))
+  end
+
+  def initialize
+    @h = Pointer(Impl).malloc(1, Impl.new(@@seed[0], @@seed[1]))
+  end
+
+  protected def initialize(a, b)
+    @h = Pointer(Impl).malloc(1, Impl.new(a, b))
+  end
+
+  def self.build
+    h = Impl.new(@@seed[0], @@seed[1])
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def self.hashit(v)
+    build do |s|
+      s << v
+    end
+  end
+
+  protected def self.build(a, b)
+    h = Impl.new(a, b)
+    s = new(pointerof(h))
+    yield s
+    s.digest
+  end
+
+  def clone_build
+    self.class.build(@h.value.a, @h.value.b) do |s|
+      yield s
+    end
+  end
+
+  def clone
+    self.class.new(@h.value.a, @h.value.b)
+  end
+
+  def <<(v : Nil)
+    @h.value.permute_nil(@@seed[2])
+    self
+  end
+
+  # mix raw value without number normalizing
+  def raw(v : Int8 | Int16 | Int32 | UInt8 | UInt16 | UInt32)
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  # mix raw value without number normalizing
+  def raw(v : Int64 | UInt64)
+    @h.value.permute((v >> 32).to_u32, @@seed[2])
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  def <<(v : Int8 | Int16 | UInt8 | UInt16)
+    @h.value.permute(v.to_u32, @@seed[2])
+    self
+  end
+
+  def <<(b : Bytes)
+    @h.value.permute(b, @@seed[2])
+    self
+  end
+
+  def <<(v)
+    v.hash(self)
+    self
+  end
+
+  def remix(v : Value)
+    self << Raw32.new(v)
+  end
+
+  def digest
+    @h.value.digest(@@seed[3])
+  end
+
+  private struct Impl
+    getter a : UInt32 = 0_u32
+    getter b : UInt32 = 0_u32
+
+    def initialize(@a : UInt32, @b : UInt32)
+    end
+
+    def permute(v : UInt32, s : UInt32)
+      permute(v, s, pointerof(@a), pointerof(@b))
+    end
+
+    def permute_nil(s : UInt32)
+      @a += s | 1
+      # LFSR
+      mx = (@b.to_i32 >> 31).to_u32 & 0xa8888eef_u32
+      @b = (@b << 1) ^ mx
+    end
+
+    def digest(seed)
+      a, b = @a, @b
```

Comment:
I want `digest` to not mutate hasher. So, that is not optimisation, but semantic.